src/x/xfullscreen.c

```diff
@@ -922,6 +922,13 @@ bool _al_xglx_get_monitor_info(ALLEGRO_SYSTEM_XGLX *s, int adapter, ALLEGRO_MONI
       info->y1 = 0;
       info->x2 = DisplayWidth(s->x11display, DefaultScreen(s->x11display));
       info->y2 = DisplayHeight(s->x11display, DefaultScreen(s->x11display));
+      int x2_mm = DisplayWidthMM(s->x11display, DefaultScreen(s->x11display));
+      int y2_mm = DisplayHeightMM(s->x11display, DefaultScreen(s->x11display));
+#define INCHES_PER_MM 0.039370
+      int dpi_hori = (info->x2 - info->x1) / (INCHES_PER_MM * x2_mm);
+      int dpi_vert = (info->y2 - info->y1) / (INCHES_PER_MM * y2_mm);
+#undef INCHES_PER_MM
+      info->dpi = sqrt(dpi_hori * dpi_vert);
```

Comment:
maybe we should report both horizontal and vertical dpi like X11?

Reply:
I was thinking just returning display dimensions, and let the users calculate DPI themselves. But then it's also more convenient to have it ready.

X11 doesn't return DPI per se, but display dimensions in millimeters.

Reply:
How would the OS know how big the monitor is?

Reply:
That's a good question! I believe there's some driver magic going on under the hood to guess or find out the display dimensions. I remember Xorg used to let you configure them manually, so maybe they are not always so reliable after all.

Reply:
I believe it'd help to read this: https://www.x.org/wiki/Development/Documentation/HowVideoCardsWork/

Reply:
Asking around, it turns out there is,

https://en.wikipedia.org/wiki/Extended_Display_Identification_Data

And in Linux kernel there's DRM and EDID with some drivers implementing it. I'd say for external displays it may not be reliable or in most cases possible at all, but laptops, embedded displays should be fine.